From c57406eb154d2ccf6758403bf120712f03ca76f3 Mon Sep 17 00:00:00 2001
From: Sourajit Karmakar <sourajit.karmakar@gmail.com>
Date: Tue, 29 Dec 2020 21:50:09 +0100
Subject: [PATCH 7/8] SystemUI: Add delete action to screen recorder
 notification

This reverts commit 6fcdb6bf9e396f89cd497ee82811dbef6179caba.

Change-Id: I125d2aff9b406341a533cbfe80686ee473dd3e88
Signed-off-by: PainKiller3 <ninadpatil100@gmail.com>
Reviewed-on: https://review.statixos.com/c/android_frameworks_base/+/6756
Reviewed-by: Anay Wadhera <anay1018@gmail.com>
Tested-by: Anay Wadhera <anay1018@gmail.com>
Reviewed-on: https://review.statixos.com/c/android_frameworks_base/+/7199
---
 packages/SystemUI/res/values/strings.xml      |  4 +++
 .../screenrecord/RecordingService.java        | 35 +++++++++++++++++++
 2 files changed, 39 insertions(+)

diff --git a/packages/SystemUI/res/values/strings.xml b/packages/SystemUI/res/values/strings.xml
index 977adde6..a9dd7796 100644
--- a/packages/SystemUI/res/values/strings.xml
+++ b/packages/SystemUI/res/values/strings.xml
@@ -285,11 +285,15 @@
     <string name="screenrecord_stop_label">Stop</string>
     <!-- Label for notification action to share screen recording [CHAR LIMIT=35] -->
     <string name="screenrecord_share_label">Share</string>
+    <!-- Label for notification action to delete a screen recording file [CHAR LIMIT=35] -->
+    <string name="screenrecord_delete_label">Delete</string>
     <!-- A toast message shown after successfully canceling a screen recording [CHAR LIMIT=NONE] -->
     <!-- Notification text shown after saving a screen recording [CHAR LIMIT=100] -->
     <string name="screenrecord_save_title">Screen recording saved</string>
     <!-- Subtext for a notification shown after saving a screen recording to prompt the user to view it [CHAR_LIMIT=100] -->
     <string name="screenrecord_save_text">Tap to view</string>
+    <!-- A toast message shown after successfully deleting a screen recording [CHAR LIMIT=NONE] -->
+    <string name="screenrecord_delete_description">Screen recording deleted</string>
     <!-- A toast message shown when there is an error deleting a screen recording [CHAR LIMIT=NONE] -->
     <string name="screenrecord_delete_error">Error deleting screen recording</string>
     <!-- A toast message shown when the screen recording cannot be started due to insufficient permissions [CHAR LIMIT=NONE] -->
diff --git a/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java b/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
index 04776266..885e88a6 100644
--- a/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
+++ b/packages/SystemUI/src/com/android/systemui/screenrecord/RecordingService.java
@@ -22,6 +22,7 @@ import android.app.NotificationChannel;
 import android.app.NotificationManager;
 import android.app.PendingIntent;
 import android.app.Service;
+import android.content.ContentResolver;
 import android.content.Context;
 import android.content.Intent;
 import android.content.res.Resources;
@@ -76,6 +77,7 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
     private static final String ACTION_STOP_NOTIF =
             "com.android.systemui.screenrecord.STOP_FROM_NOTIF";
     private static final String ACTION_SHARE = "com.android.systemui.screenrecord.SHARE";
+    private static final String ACTION_DELETE = "com.android.systemui.screenrecord.DELETE";
     private static final String PERMISSION_SELF = "com.android.systemui.permission.SELF";
 
     private final RecordingController mController;
@@ -207,6 +209,23 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 // Close quick shade
                 sendBroadcast(new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS));
                 break;
+            case ACTION_DELETE:
+                // Close quick shade
+                sendBroadcast(new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS));
+
+                ContentResolver resolver = getContentResolver();
+                Uri uri = Uri.parse(intent.getStringExtra(EXTRA_PATH));
+                resolver.delete(uri, null, null);
+
+                Toast.makeText(
+                        this,
+                        R.string.screenrecord_delete_description,
+                        Toast.LENGTH_LONG).show();
+
+                // Remove notification
+                mNotificationManager.cancelAsUser(null, NOTIFICATION_VIEW_ID, currentUser);
+                Log.d(TAG, "Deleted recording " + uri);
+                break;
         }
         return Service.START_STICKY;
     }
@@ -363,6 +382,16 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                         PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE))
                 .build();
 
+        Notification.Action deleteAction = new Notification.Action.Builder(
+                Icon.createWithResource(this, R.drawable.ic_screenrecord),
+                getResources().getString(R.string.screenrecord_delete_label),
+                PendingIntent.getService(
+                        this,
+                        REQUEST_CODE,
+                        getDeleteIntent(this, uri.toString()),
+                        PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE))
+                .build();
+
         Bundle extras = new Bundle();
         extras.putString(Notification.EXTRA_SUBSTITUTE_APP_NAME,
                 getResources().getString(R.string.screenrecord_name));
@@ -377,6 +406,7 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                         viewIntent,
                         PendingIntent.FLAG_IMMUTABLE))
                 .addAction(shareAction)
+                .addAction(deleteAction)
                 .setAutoCancel(true)
                 .addExtras(extras);
 
@@ -478,6 +508,11 @@ public class RecordingService extends Service implements ScreenMediaRecorderList
                 .putExtra(EXTRA_PATH, path);
     }
 
+    private static Intent getDeleteIntent(Context context, String path) {
+        return new Intent(context, RecordingService.class).setAction(ACTION_DELETE)
+                .putExtra(EXTRA_PATH, path);
+    }
+
     @Override
     public void onInfo(MediaRecorder mr, int what, int extra) {
         Log.d(TAG, "Media recorder info: " + what);
-- 
2.40.0

